src/hazelcore/UHExp.re

```diff
@@ -305,10 +305,119 @@ let text_operand =
     );
   };
 
-let associate = (seq: seq) => {
-  let skel_str = Skel.mk_skel_str(seq, Operators_Exp.to_parse_string);
-  let lexbuf = Lexing.from_string(skel_str);
-  SkelExprParser.skel_expr(SkelExprLexer.read, lexbuf);
+let associate = (seq: seq): skel => {
+  /**
+   * Shunting-yard opp algorithm according to the specification here
+   * https://en.wikipedia.org/wiki/Shunting-yard_algorithm#The_algorithm_in_detail
+   *
+   */
+  let rec go_seq =
+          (
+            skels: list(skel), /* List of skels to be combined into single output skel. */
+            op_stack: list(Operators_Exp.t), /* Stack of operators. */
+            seq: seq, /* Convert this seq to output skel. */
+            lex_addr: int,
+          ) /* Lexical address of the current operand. */
+          : skel => {
+    switch (seq) {
+    | S(_, affix) =>
+      /**
+       * If the next token is an operand, add a new operand to the output skel.
+       */
+      go_affix(
+        [Skel.Placeholder(lex_addr), ...skels],
+        op_stack,
+        affix, /* Tail of seq without first operand. */
+        lex_addr + 1 /* Increment lexical address of operand in seq. */
+      )
+    };
+  }
+  and go_affix =
+      (
+        skels: list(skel),
+        op_stack: list(Operators_Exp.t),
+        affix: affix,
+        lex_addr: int,
+      )
+      : skel => {
+    switch (affix) {
+    | A(op, seq) =>
+      /**
+       * If the next token is an operator, pop operators from the operator stack,
+       * then push this operator to the operator stack.
+       */
+      let should_mv = op' =>
+        /**
+         * Continue popping operators while the precedence of the top of the operator
+         * stack has greater precedence than the current operator.
+         */
+        {
+          switch (Operators_Exp.associativity(op')) {
+          | Left =>
+            Operators_Exp.precedence(op) <= Operators_Exp.precedence(op')
```

Comment:
@cyrus- This logic. Also sounds fair, I'll setup some tests and look into perf.

Reply:
ah gotcha. maybe mention explicitly what's going on here in comments.